diff --git a/src/core/ngx_connection.c b/src/core/ngx_connection.c
index 9afb952..e9c9007 100644
--- a/src/core/ngx_connection.c
+++ b/src/core/ngx_connection.c
@@ -550,6 +550,9 @@ ngx_get_connection(ngx_socket_t s, ngx_log_t *log)
     ngx_uint_t         instance;
     ngx_event_t       *rev, *wev;
     ngx_connection_t  *c;
+#if (NGX_STAT_EXTENDED)
+    conn_score        *cs;
+#endif
 
     /* disable warning: Win32 SOCKET is u_int while UNIX socket is int */
 
@@ -587,8 +590,16 @@ ngx_get_connection(ngx_socket_t s, ngx_log_t *log)
     rev = c->read;
     wev = c->write;
 
+#if (NGX_STAT_EXTENDED)
+    cs = c->cs;
+#endif
+
     ngx_memzero(c, sizeof(ngx_connection_t));
 
+#if (NGX_STAT_EXTENDED)
+    c->cs = cs;
+#endif
+
     c->read = rev;
     c->write = wev;
     c->fd = s;
diff --git a/src/core/ngx_connection.h b/src/core/ngx_connection.h
index 1694dd4..b3ee586 100644
--- a/src/core/ngx_connection.h
+++ b/src/core/ngx_connection.h
@@ -97,6 +97,11 @@ typedef enum {
 
 
 struct ngx_connection_s {
+
+#if (NGX_STAT_EXTENDED)
+    void               *cs;
+#endif
+
     void               *data;
     ngx_event_t        *read;
     ngx_event_t        *write;
diff --git a/src/event/ngx_event.c b/src/event/ngx_event.c
index 563186e..bafc3ff 100644
--- a/src/event/ngx_event.c
+++ b/src/event/ngx_event.c
@@ -57,7 +57,7 @@ ngx_int_t             ngx_accept_disabled;
 ngx_file_t            ngx_accept_mutex_lock_file;
 
 
-#if (NGX_STAT_STUB)
+#if (NGX_STAT_STUB || NGX_STAT_EXTENDED)
 
 ngx_atomic_t   ngx_stat_accepted0;
 ngx_atomic_t  *ngx_stat_accepted = &ngx_stat_accepted0;
@@ -74,7 +74,15 @@ ngx_atomic_t  *ngx_stat_writing = &ngx_stat_writing0;
 
 #endif
 
+#if (NGX_STAT_EXTENDED)
 
+ngx_uint_t      ngx_num_workers;
+size_t          shm_size = 0;
+worker_score   *workers;
+conn_score     *conns;
+worker_score   *my_worker;
+
+#endif
 
 static ngx_command_t  ngx_events_commands[] = {
 
@@ -342,6 +350,40 @@ ngx_handle_read_event(ngx_event_t *rev, u_int flags)
     return NGX_OK;
 }
 
+#if (NGX_STAT_EXTENDED)
+
+static void
+init_request_cnt(request_cnt *recent_request_cnt)
+{
+    ngx_uint_t  i;
+
+    for (i = 0; i < RECENT_PERIOD; i++) {
+        recent_request_cnt[i].time = 0;
+        recent_request_cnt[i].cnt = 0;
+    }
+}
+
+
+static void
+init_workers(worker_score *workers)
+{
+    worker_score  *ws;
+    ngx_uint_t  i;
+
+    for (i = 0; i < ngx_num_workers; i++)
+    {
+        ws = (worker_score *)((char *) workers + WORKER_SCORE_LEN * i);
+
+        ws->pid = 0;
+        ws->access_count = 0;
+        ws->bytes_sent = 0;
+        ws->mode = SERVER_READY;
+
+        init_request_cnt(ws->recent_request_cnt);
+    }
+}
+
+#endif
 
 ngx_int_t
 ngx_handle_write_event(ngx_event_t *wev, size_t lowat)
@@ -484,7 +526,13 @@ ngx_event_module_init(ngx_cycle_t *cycle)
     }
 
     if (ngx_accept_mutex_ptr) {
+#if (NGX_STAT_EXTENDED)
+        shm.addr = (u_char *) ngx_accept_mutex_ptr;
+        shm.size = shm_size;
+        ngx_shm_free(&shm);
+#else
         return NGX_OK;
+#endif
     }
 
 
@@ -495,7 +543,7 @@ ngx_event_module_init(ngx_cycle_t *cycle)
     size = cl            /* ngx_accept_mutex */
            + cl;         /* ngx_connection_counter */
 
-#if (NGX_STAT_STUB)
+#if (NGX_STAT_STUB || NGX_STAT_EXTENDED)
 
     size += cl           /* ngx_stat_accepted */
            + cl          /* ngx_stat_handled */
@@ -506,6 +554,20 @@ ngx_event_module_init(ngx_cycle_t *cycle)
 
 #endif
 
+#if (NGX_STAT_EXTENDED)
+
+    if (WORKER_SCORE_LEN < sizeof(worker_score)) {
+        ngx_log_error(NGX_LOG_EMERG, cycle->log, 0, "worker_score(%d) too large", sizeof(worker_score ));
+        return NGX_ERROR;
+    }
+
+    size += (WORKER_SCORE_LEN * ccf->worker_processes)
+        + (sizeof(conn_score) * ccf->worker_processes * cycle->connection_n);
+
+    shm_size = size;
+
+#endif
+
     shm.size = size;
     shm.log = cycle->log;
 
@@ -525,7 +587,7 @@ ngx_event_module_init(ngx_cycle_t *cycle)
 
     ngx_connection_counter = (ngx_atomic_t *) (shared + 1 * cl);
 
-#if (NGX_STAT_STUB)
+#if (NGX_STAT_STUB || NGX_STAT_EXTENDED)
 
     ngx_stat_accepted = (ngx_atomic_t *) (shared + 2 * cl);
     ngx_stat_handled = (ngx_atomic_t *) (shared + 3 * cl);
@@ -1083,6 +1145,17 @@ ngx_event_debug_connection(ngx_conf_t *cf, ngx_command_t *cmd, void *conf)
 
 #endif
 
+#if (NGX_STAT_EXTENDED)
+
+    ngx_num_workers = ccf->worker_processes;
+
+    workers = (worker_score *) (shared + 10 * cl);
+    init_workers(workers);
+
+    conns = (conn_score *) (shared + 10 * cl + ccf->worker_processes * WORKER_SCORE_LEN);
+
+#endif
+
     return NGX_OK;
 }
 
diff --git a/src/event/ngx_event.h b/src/event/ngx_event.h
index e1388d5..8e35cf7 100644
--- a/src/event/ngx_event.h
+++ b/src/event/ngx_event.h
@@ -7,7 +7,9 @@
 #ifndef _NGX_EVENT_H_INCLUDED_
 #define _NGX_EVENT_H_INCLUDED_
 
-
+#if (NGX_STAT_EXTENDED)
+#include <sys/times.h>
+#endif
 #include <ngx_config.h>
 #include <ngx_core.h>
 
@@ -455,7 +457,7 @@ extern ngx_msec_t             ngx_accept_mutex_delay;
 extern ngx_int_t              ngx_accept_disabled;
 
 
-#if (NGX_STAT_STUB)
+#if (NGX_STAT_STUB || NGX_STAT_EXTENDED)
 
 extern ngx_atomic_t  *ngx_stat_accepted;
 extern ngx_atomic_t  *ngx_stat_handled;
@@ -466,6 +468,87 @@ extern ngx_atomic_t  *ngx_stat_writing;
 
 #endif
 
+#if (NGX_STAT_EXTENDED)
+
+#define  SCORE__REQUEST_LEN            128
+#define  SCORE__CLIENT_LEN              32
+#define  SCORE__VHOST_LEN               32
+
+#define  CL_SIZE                       128
+#define  WORKER_SCORE_LEN              640      /* 128 X 5 */
+
+#define  SERVER_READY                   '-'
+#define  SERVER_BUSY_READ               'R'
+#define  SERVER_BUSY_WRITE              'W'
+#define  SERVER_BUSY_LOG                'L'
+#define  SERVER_DEAD                    'I'
+
+#define  RECENT_PERIOD                  64
+#define  RECENT_MASK            0x0000003F
+
+typedef struct {
+    uint32_t  time;
+    uint32_t  cnt;
+} request_cnt;
+
+
+typedef struct {
+    request_cnt     recent_request_cnt[RECENT_PERIOD];
+    struct  tms     times;
+    time_t          last_used;
+    ngx_uint_t      bytes_sent;
+    ngx_uint_t      access_count;
+    ngx_pid_t       pid;
+    char            mode;
+} worker_score;
+
+
+typedef struct {
+    u_char  request[SCORE__REQUEST_LEN];
+    u_char  client[SCORE__CLIENT_LEN];
+    u_char  vhost[SCORE__VHOST_LEN];
+    time_t          last_used;
+    ngx_uint_t      bytes_sent;
+    ngx_uint_t      access_count;
+    char            mode ;
+    ngx_msec_int_t  response_time;
+    ngx_msec_int_t  upstream_response_time;
+    ngx_uint_t      status;
+    size_t          zin;
+    size_t          zout;
+    char            active;
+} conn_score;
+
+
+extern  worker_score  *workers;
+extern  worker_score  *my_worker;
+extern  conn_score    *conns;
+
+
+static inline void
+set_conn_active(ngx_connection_t *c)
+{
+    conn_score  *score = (conn_score *) c->cs;
+    score->active = 1 ;
+}
+
+
+static inline void
+set_conn_inactive(ngx_connection_t *c)
+{
+    conn_score  *score = (conn_score *) c->cs;
+    score->active = 0 ;
+}
+
+
+static inline void
+set_conn_mode(ngx_connection_t *c, char mode)
+{
+    conn_score  *score = (conn_score *) c->cs;
+    score->mode = mode ;
+}
+
+#endif
 
 #define NGX_UPDATE_TIME         1
 #define NGX_POST_EVENTS         2
diff --git a/src/http/modules/ngx_http_gzip_filter_module.c b/src/http/modules/ngx_http_gzip_filter_module.c
index 7210dd6..fafe46c 100644
--- a/src/http/modules/ngx_http_gzip_filter_module.c
+++ b/src/http/modules/ngx_http_gzip_filter_module.c
@@ -970,6 +970,15 @@ ngx_http_gzip_ratio_variable(ngx_http_request_t *r,
 
     v->len = ngx_sprintf(v->data, "%ui.%02ui", zint, zfrac) - v->data;
 
+#if (NGX_STAT_EXTENDED)
+    {
+        conn_score  *cs = r->connection->cs;
+        
+        cs->zin = ctx->zin;
+        cs->zout = ctx->zout;
+    }
+#endif
+
     return NGX_OK;
 }
 
diff --git a/src/http/ngx_http_request.c b/src/http/ngx_http_request.c
index 2c1ac95..368eae2 100644
--- a/src/http/ngx_http_request.c
+++ b/src/http/ngx_http_request.c
@@ -2345,6 +2345,60 @@ ngx_http_request_done(ngx_http_request_t *r, ngx_int_t error)
     ngx_destroy_pool(r->pool);
 }
 
+#if (NGX_STAT_EXTENDED)
+
+static inline void
+update_request_cnt(worker_score *worker, conn_score *cs, uint32_t sec)
+{
+    uint32_t  index;
+
+    index = sec & RECENT_MASK;
+    if (sec == worker->recent_request_cnt[index].time) {
+        worker->recent_request_cnt[index].cnt += 1;
+    }
+    else {
+        worker->recent_request_cnt[index].time = sec;
+        worker->recent_request_cnt[index].cnt = 1;
+    }
+}
+
+
+static inline  ngx_msec_int_t
+get_response_time(ngx_time_t *tp, ngx_http_request_t *r)
+{
+    ngx_msec_int_t  ms;
+
+    ms = (ngx_msec_int_t) ((tp->sec - r->start_sec) * 1000 + (tp->msec - r->start_msec));
+    ms = ( 0 <= ms ) ? ms : 0;
+
+    return ms;
+}
+
+static inline  ngx_msec_int_t
+get_proxy_response_time(ngx_http_request_t  *r)
+{
+    ngx_http_upstream_state_t  *state;
+    ngx_msec_int_t  ms = -1;
+    ngx_uint_t  i;
+
+    if (NULL != r->upstream_states && 0 < r->upstream_states->nelts)
+    {
+        state = r->upstream_states->elts;
+        for (i = 0; i < r->upstream_states->nelts; i++)
+        {
+            if (0 != state[i].status)
+            {
+                ms = (ngx_msec_int_t) (state[i].response_sec * 1000 + state [i].response_msec);
+                ms = (0 <= ms) ? ms : 0;
+                break;
+            }
+        }
+    }
+
+    return ms;
+}
+
+#endif
 
 static void
 ngx_http_close_connection(ngx_connection_t *c)
@@ -2376,6 +2430,8 @@ ngx_http_close_connection(ngx_connection_t *c)
     ngx_close_connection(c);
 
     ngx_destroy_pool(pool);
+
+
 }
 
 
diff --git a/src/http/ngx_http_request.h b/src/http/ngx_http_request.h
index 9df5bd2..e50d46a 100644
--- a/src/http/ngx_http_request.h
+++ b/src/http/ngx_http_request.h
@@ -466,7 +466,7 @@ struct ngx_http_request_s {
     unsigned                          filter_need_temporary:1;
     unsigned                          allow_ranges:1;
 
-#if (NGX_STAT_STUB)
+#if (NGX_STAT_STUB || NGX_STAT_EXTENDED)
     unsigned                          stat_reading:1;
     unsigned                          stat_writing:1;
 #endif
diff --git a/src/os/unix/ngx_process.c b/src/os/unix/ngx_process.c
index a1454b6..421dd51 100644
--- a/src/os/unix/ngx_process.c
+++ b/src/os/unix/ngx_process.c
@@ -32,6 +32,10 @@ ngx_socket_t     ngx_channel;
 ngx_int_t        ngx_last_process;
 ngx_process_t    ngx_processes[NGX_MAX_PROCESSES];
 
+#if (NGX_STAT_EXTENDED)
+ngx_int_t        old_ngx_last_process = 0;
+ngx_process_t    old_ngx_processes[NGX_MAX_PROCESSES];
+#endif
 
 ngx_signal_t  signals[] = {
     { ngx_signal_value(NGX_RECONFIGURE_SIGNAL),
diff --git a/src/os/unix/ngx_process.h b/src/os/unix/ngx_process.h
index a34ca61..e926389 100644
--- a/src/os/unix/ngx_process.h
+++ b/src/os/unix/ngx_process.h
@@ -81,5 +81,9 @@ extern ngx_int_t      ngx_process_slot;
 extern ngx_int_t      ngx_last_process;
 extern ngx_process_t  ngx_processes[NGX_MAX_PROCESSES];
 
+#if (NGX_STAT_EXTENDED)
+extern ngx_int_t      old_ngx_last_process;
+extern ngx_process_t  old_ngx_processes[NGX_MAX_PROCESSES];
+#endif
 
 #endif /* _NGX_PROCESS_H_INCLUDED_ */
diff --git a/src/os/unix/ngx_process_cycle.c b/src/os/unix/ngx_process_cycle.c
index a262ba5..b26426a 100644
--- a/src/os/unix/ngx_process_cycle.c
+++ b/src/os/unix/ngx_process_cycle.c
@@ -207,6 +207,15 @@ ngx_master_process_cycle(ngx_cycle_t *cycle)
 
             ngx_log_error(NGX_LOG_NOTICE, cycle->log, 0, "reconfiguring");
 
+#if (NGX_STAT_EXTENDED)
+            old_ngx_last_process = ngx_last_process;
+            for (i = 0; i < old_ngx_last_process; i++)
+                old_ngx_processes[i] = ngx_processes[i];
+
+            ngx_last_process = 0;
+            ngx_process_slot = 0;
+#endif
+
             cycle = ngx_init_cycle(cycle);
             if (cycle == NULL) {
                 cycle = (ngx_cycle_t *) ngx_cycle;
@@ -424,6 +433,64 @@ ngx_signal_worker_processes(ngx_cycle_t *cycle, int signo)
 
     ch.fd = -1;
 
+#if (NGX_STAT_EXTENDED)
+    if ( NGX_CMD_QUIT == ch.command && 0 < old_ngx_last_process )
+    {
+        for (i = 0; i < old_ngx_last_process; i++) {
+            ngx_log_debug7(NGX_LOG_DEBUG_EVENT, cycle->log, 0,
+                           "child: %d %P e:%d t:%d d:%d r:%d j:%d",
+                           i,
+                           old_ngx_processes[i].pid,
+                           old_ngx_processes[i].exiting,
+                           old_ngx_processes[i].exited,
+                           old_ngx_processes[i].detached,
+                           old_ngx_processes[i].respawn,
+                           old_ngx_processes[i].just_spawn);
+            
+            if (old_ngx_processes[i].detached || old_ngx_processes[i].pid == -1)
+                continue;
+            
+            if (old_ngx_processes[i].exiting && signo == ngx_signal_value( NGX_SHUTDOWN_SIGNAL))
+                continue;
+            
+            if (ch.command) {
+                if (ngx_write_channel(old_ngx_processes[i].channel[0], &ch, sizeof(ngx_channel_t), cycle->log) == NGX_OK) {
+                    if (signo != ngx_signal_value(NGX_REOPEN_SIGNAL))
+                        old_ngx_processes[i].exiting = 1;
+                    
+                    continue;
+                }
+            }
+            
+            ngx_log_debug2(NGX_LOG_DEBUG_CORE, cycle->log, 0,
+                           "kill (%P, %d)", old_ngx_processes[i].pid, signo);
+            
+            if (kill(old_ngx_processes[i].pid, signo) == -1) {
+                err = ngx_errno;
+                ngx_log_error(NGX_LOG_ALERT, cycle->log, err,
+                              "kill(%P, %d) failed", old_ngx_processes[i].pid, signo);
+                
+                if (err == NGX_ESRCH) {
+                    old_ngx_processes[i].exited = 1;
+                    old_ngx_processes[i].exiting = 0;
+                    ngx_reap = 1;
+                }
+                continue;
+            }
+            
+            if (signo != ngx_signal_value(NGX_REOPEN_SIGNAL))
+                old_ngx_processes[i].exiting = 1;
+        }
+        
+        old_ngx_last_process = -1;
+
+        for (i = 0; i < ngx_last_process; i++) {
+            ngx_processes [i].just_spawn = 0 ;
+        }
+     
+        return ;
+    }
+#endif
 
     for (i = 0; i < ngx_last_process; i++) {
 
@@ -673,6 +740,21 @@ ngx_worker_process_cycle(ngx_cycle_t *cycle, void *data)
 
     ngx_setproctitle("worker process");
 
+#if (NGX_STAT_EXTENDED)
+    {
+        my_worker = (worker_score *) ((char *) workers + WORKER_SCORE_LEN * ngx_process_slot);
+        ngx_time_t  *tp = ngx_timeofday();
+        
+        my_worker->last_used = tp->sec;
+        my_worker->pid = ngx_getpid();
+        my_worker->bytes_sent = 0L;
+        my_worker->access_count = 0L;
+        my_worker->mode = SERVER_READY;
+        
+        times(&my_worker->times);
+    }
+#endif
+
 #if (NGX_THREADS)
 
     ccf = (ngx_core_conf_t *) ngx_get_conf(cycle->conf_ctx, ngx_core_module);
